<!DOCTYPE html>
<html lang="en">

<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Chapter 9 PS</title>
</head>

<body>
    <script>
        /* Q1. Write a progrm to load a Javascript file in a browser using Promises. Use .then() to display
        an alert when the load is complete. */

        // const loadScript =(src)=>{
        //     return new Promise((resolve, reject) => {
        //     let script = document.createElement("script")
        //     document.head.append(script)
        //     script.src = src;
        //     script.onload = ()=>{
        //         resolve(src + " Done Success!");
        //     }
        // })
        // }
        //   let a = loadScript("https://cdn.jsdelivr.net/npm/bootstrap@5.2.1/dist/js/bootstrap.bundle.min.js")
        //   a.then((value)=>{
        //     alert(value)
        //   })

        /* Q2. Write the same program from previous question and use async/await syntax. */

        //   const loadScript = async (src)=>{
        //     return new Promise((resolve, reject) => {
        //     let script = document.createElement("script")
        //     document.head.append(script)
        //     script.src = src;
        //     script.onload = ()=>{
        //         resolve(src + " Done Success!");
        //     }
        // })
        // }
        //     let run = async()=>{

        //         console.log(new Date().getMilliseconds());
        //         let a = await loadScript("https://cdn.jsdelivr.net/npm/bootstrap@5.2.1/dist/js/bootstrap.bundle.min.js")
        //         console.log(a);
        //         console.log(new Date().getMilliseconds());
        //     }
        //     run();

        /* Q3. Create a Promise which rejects after 3 seconds. Use an async/await to get it's value. Use a
        try catch to handle it's error. */

        // let p = () => {
        //     return new Promise((resolve, reject) => {
        //         setTimeout(() => {
        //             reject("This is not acceptable.")
        //         }, 3000)
        //     })
        // }
        // let a = async ()=>{
        //     try {
        //         let c = await p();
        //         console.log(c);
        //     } catch (error) {
        //         console.log("Error is handled");
        //     }
        // }
        // a();

        /* Q4. Write a Program using Promise.all() inside an async/await to await 3 promises. Compare it's results with the case where we await the promises one by one. */

        let p1 = async()=>{
            return new Promise((resolve, reject) => {
                setTimeout(()=>{
                    resolve(30);
                }, 1000)
            })
        } 
        let p2 = async()=>{
            return new Promise((resolve, reject) => {
                setTimeout(()=>{
                    resolve(40);
                }, 3000)
            })
        } 
        let p3 = async()=>{
            return new Promise((resolve, reject) => {
                setTimeout(()=>{
                    resolve(50);
                }, 2000)
            })
        } 

        let a = async()=>{
            // console.time("a");
            // let a1 = await p1();
            // let a2 = await p2();
            // let a3 = await p3();
            // console.log(a1,a2,a3);
            // console.timeEnd("a");

            console.time("a");
            let a1a2a3 = await Promise.all([p1(),p2(),p3()])   // --> This one is used for optimization.
            console.log(a1a2a3);
            console.timeEnd("a");
        }
        a();
    </script>
</body>

</html>